### ISSUE TITLE
Improve accessibility (alt tags, etc)

### ISSUE THREAD

**Issue by taylor-elam** [grafana/grafana#5986]
- **I'm submitting a ...**
- [x] Bug report
- [ ] Feature request

**Bug description:**
All images must have alternate text to convey their purpose and meaning to screen reader users.

**Reproduce the issue:**
This FireFox plugin will allow you to test accessibility
https://addons.mozilla.org/en-us/firefox/addon/axe-devtools/

**Solution**
Ensure all informative `<img>` elements have short, descriptive alternate text and all decorative `<img>` elements have empty alt attributes (e.g. `alt=""`).

**Version**
Grafana: 2.6.0

**Comment by bergquist** [comment]
@taylor-elam Just curious. Do you have any experience with people using screen readers for grafana? Since its 90% about graphs.

**Comment by torkelo** [comment]
do you have many users who use screen readers with Grafana? Grafana is a visualization tool so accessibility for visually impaired is not a big concert right now.

**Comment by david-kirkland** [comment]
I just wanted to jump in and state that data visualization can be made ADA compliant.  Its not always easy or straightforward, but it can be done.  In this case it would be very simple to add an alt tag to the Grafana logo.  If I'm not mistaken, having an alt attribute on an image is required for the HTML to be valid anyways.  I'm not sure I understand why this issue was closed.  It's a super simple fix.

With all that said, some enterprise applications have ADA requirements which extend to the 3rd party code included in the app.  Having non-compliant 3rd party code (like Grafana) would mean that the app can not be certified fully compliant with WCAG 2.0.  This may incentivize enterprises to seek alternatives to Grafana.  Just food for thought.

**Comment by torkelo** [comment]
how would an alt tag on the Grafana logo help a visualization tool like Grafana to be easier to use for the visually impaired? 

But I see your point. I actually got an email from some government agency asking about the same thing. But I do not understand the details of this request. If it's only about alt tags it seems like just a token thing to meet the ADA requirements, nothing substantial that actually makes the software usable for someone who is blind/partially blind. For example what about the hundreds of icons (they do not use image tag).

**Comment by david-kirkland** [comment]
Government agencies aren't supposed to use anything that is not ADA compliant.  They are more strict on that than enterprises.

There are two types of images from an ADA perspective; decorative and non-decorative.

Decorative images are the type that do not convey important information and are usually just there for visually aesthetic purposes.  You can tell a screen reader to ignore these images by adding an empty alt tag, like so... `alt=""`.  This basically makes them invisible to a screen reader.

Non-decorative images should always include an `alt` attribute with a meaningful value.  Without an `alt` attribute the screen reader will just say "image" when it comes to it.  This leaves the screen reader user wondering what information the image might convey to non-screen reader users.

In this case, the image is in a link without any text in the link.  The screen reader user will know that this is a clickable image, but will have no idea what is going to happen when they click it.  Without this information, the screen reader user will not be able to effectively navigate the application.  Non-screen reader users get all this information visually.

For this situation I would recommend...

1. Add `href=""` to the links.  Anchors without a `href` attribute are not valid, so the screen reader skips it which means screen reader users won't even "see" this link.
2. Add `alt="Grafana logo"` to the image.s
4. Add screen reader only text to the links with Bootstrap's `sr-only` class.

```
<a href="" class="pointer sidemenu-top-btn" ng-click="contextSrv.toggleSideMenu()">
  <img class="logo-icon" src="img/fav32.png" alt="Grafana logo">
  <i class="pull-right fa fa-angle-left"></i>
  <span class="sr-only">Close Menu</span>
</a>

...

<a class="pointer top-nav-menu-btn" ng-if="!contextSrv.sidemenu" ng-click="contextSrv.toggleSideMenu()">
  <img class="logo-icon" src="img/fav32.png" alt="Grafana logo">
  <i class="fa fa-bars"></i>
  <span class="sr-only">Open Menu</span>
</a>
```

If your version of Bootstrap doesn't have the `sr-only` class, add it...

```
.sr-only {
    position: absolute;
    width: 1px;
    height: 1px;
    margin: -1px;
    padding: 0;
    overflow: hidden;
    clip: rect(0,0,0,0);
    border: 0;
}
```

Since in this case the logo is just a placeholder for the link, which is what really matters, you could replace `alt="Grafana logo"` with `aria-hidden="true"` to hide the logo altogether.  In either case, the link needs text that tells the screen reader user what its for, as shown above.

**Comment by molaachir** [comment]
Hi,

I am a visually impaired developer, who used to see well enough to use grafana untill very recently. Since I switched to a screen reader the application has become virtually useless for me. There are still several things you could do to make the application usable for people with a visual impairment:
 
Expose the data in a bar chart in a table as well, make that table easily accessible through a hotkey, for example, when the image containing the bar chart has focus, you can use a hotkey to pull up a table containing the same information as the bar chart.

Other graphs are a challenge, I'm not sure yet myself on how to fix that, but a bar chart should be low hanging fruit to start with. I work in a team that uses grafana extensively, so if slowly but surely features can be added that would help me in my work without making me use all kinds of tooling that my team does not use, would be very welcome. I'll think about more improvements and post them here as I come up with them. Hope some progress can be made on this front, would be great.

**Comment by parhamdoustdar** [comment]
I'm also a completely blind developer, and I have to rely on pretty much everyone else to look at the metrics here.

One thing could also be to provide an alternative way of querying the data and presenting that in a table. That way, if I want an average, I can run a query for it, if I want averages across time ranges, I can do that, or if I want the highest/lowest values, I can query for that too.

This way this feature could be useful to visually impaired and sighted people alike – this would be one other way of looking at data.

**Comment by craigbrett17** [comment]
Happy to add my voice to this. I'm a visually impaired software dev. I've used Graffana in several jobs now, and I do like the tool, I can use my residual vision to get a feel for spikes, etc. But a tabular alternative would be absolutely great.

I know you can present a chart as a table by changing its type, but I'm not sure about changing ones that are already on a dashboard. Maybe I can try it and see

**Comment by craigbrett17** [comment]
Nope, doesn't work, unless you edit the chart.

**Comment by Neurrone** [comment]
Hoping this could also be worked on as well, as its been 4 years since the issue was first opened.

Adding the ability to view data via tables would make dashboards useful for screen reader users.

**Comment by behofer** [comment]
Hi,
I want to second this.
As a screenreader user, it's quite impossible for me to use Grafana. And that's not because it's a visualization only tool: Also Dashboards in other tools (e.g. Kibana) are really accessible and useful for me as a screenreader user.
There's a lot that can be done to make dashboards useful for visually impaired users as well - which isn't always clear for fully sighted people.

Thanks!

**Comment by codeblooded** [comment]
I would just like to bump this thread, since I am considering Grafana for testing a CNCF foundation project. The accessibility features are very important to me and my colleagues.

**Comment by blami** [comment]
I'd like to volunteer and start working on this @Ijin08

**Comment by behofer** [comment]
@blami Great! Let me know if I can be of any help (testing with screen readers, providing technical background etc.)

**Comment by AKhares** [comment]
@blami are you working on this? Just curious to know if there is a plan to get the accessibility support anytime soon.

**Comment by blami** [comment]
@AKhares yes, I am working on this in my spare time. Currently I'm adding some automated tests coverage to help me find missing accessibility markup, based on that I'll set priorities and start fixing things.

**Comment by torkelo** [comment]
To be clear I think accessibility is super important (following standards, being easy to use with keyboard, color contrasts, etc). This is a very old issue from a time when the team size was 3 people and this user was specifically asking for alt tags on images (which we have very few of, for example Grafana logo). Was not sure how that would make a difference to the overall usability of a 100% visual (at the time) tool like Grafana.

Since then we have created a squad focused on accessibility and navigation (as well as a cross org accessibility working group). Created a mission statement: https://grafana.com/accessibility/
Made lots of progress to keyboard navigation issues in v8.x and the new navigation in v9 has major accessibility improvements. To track, progress check out open or and closed accessibility issues: https://github.com/grafana/grafana/issues?q=is%3Aissue+is%3Aopen+accessibility+label%3Atype%2Faccessibility

**Comment by yaelleC** [comment]
We are tracking wider a11y epics such as : https://github.com/grafana/grafana/issues/65701, https://github.com/grafana/grafana/issues/72749 , https://github.com/grafana/grafana/issues/46474 or https://github.com/grafana/grafana/issues/79637 

We also have pa11y-ci checks for anything that can be checked automatically (such as alt-text on images)

As Torkel mentioned we now have a squad that owns accessibility, we have a VPAT and constantly work to improve the a11y within Grafana and its plugins. We are happy to take in specific bugs/feature requests related to a11y.

I will close this issue as some of it has been done, some is a duplicate of the epics mentioned above
